scanpointgenerator/core/compoundgenerator.py

```diff
@@ -21,236 +23,315 @@ def __init__(self, generators, excluders, mutators):
+            if gen_1 is gen_2:
+                continue
+            if isinstance(gen_1, LineGenerator) \
+                    and isinstance(gen_2, LineGenerator):
+                gen_1.produce_points()
+                gen_2.produce_points()
+                valid = np.full(gen_1.num, True, dtype=np.int8)
+                valid &= \
+                    gen_1.points[axis_1] <= rect.roi.width + rect.roi.start[0]
+                valid &= gen_1.points[axis_1] >= rect.roi.start[0]
+                points_1 = gen_1.points[axis_1][valid.astype(np.bool)]
+                valid = np.full(gen_2.num, True, dtype=np.int8)
+                valid &= \
+                    gen_2.points[axis_2] <= rect.roi.height + rect.roi.start[1]
+                valid &= gen_2.points[axis_2] >= rect.roi.start[1]
+                points_2 = gen_2.points[axis_2][valid.astype(np.bool)]
+                new_gen1 = LineGenerator(
+                    gen_1.name, gen_1.units, points_1[0], points_1[-1],
+                    len(points_1), gen_1.alternate_direction)
+                new_gen2 = LineGenerator(
+                    gen_2.name, gen_2.units, points_2[0], points_2[-1],
+                    len(points_2), gen_2.alternate_direction)
+                generators[generators.index(gen_1)] = new_gen1
+                generators[generators.index(gen_2)] = new_gen2
+                excluders.remove(rect)
+
+        for generator in generators:
+            generator.produce_points()
+            self.axes_points.update(generator.points)
+            self.axes_points_lower.update(generator.points_lower)
+            self.axes_points_upper.update(generator.points_upper)
+            self.num *= generator.num
 
-            # If innermost generator, use bounds
-            reverse, sub_point = self._get_sub_point(
-                num_point_sets - 1, point_num)
-            point.positions.update(sub_point.positions)
-            if reverse:  # Swap bounds if reversing
-                point.upper.update(sub_point.lower)
-                point.lower.update(sub_point.upper)
+            dim = {"size":generator.num,
+                "axes":list(generator.axes),
+                "generators":[generator],
+                "masks":[],
+                "tile":1,
+                "repeat":1,
+                "alternate":generator.alternate_direction}
+            self.dimensions.append(dim)
+
+        for excluder in excluders:
+            axis_1, axis_2 = excluder.scannables
+            # ensure axis_1 is "outer" axis (if separate generators)
+            gen_1 = [g for g in generators if axis_1 in g.axes][0]
+            gen_2 = [g for g in generators if axis_2 in g.axes][0]
+            gen_diff = generators.index(gen_1) \
+                - generators.index(gen_2)
+            if gen_diff < -1 or gen_diff > 1:
+                raise ValueError(
+                    "Excluders must be defined on axes that are adjacent in " \
```

Comment:
It's not a problem at present, we can revisit it if it's a problem in the future...